massa-execution/src/worker.rs

```diff
@@ -48,52 +55,78 @@ pub struct ExecutionWorker {
     ordered_active_blocks: Vec<(BlockId, Block)>,
     /// pending CSS final blocks
     ordered_pending_css_final_blocks: Vec<(BlockId, Block)>,
-    /// VM
+    /// VM thread
     vm_thread: JoinHandle<()>,
     /// VM execution requests queue
     execution_queue: ExecutionQueue,
 }
 
 impl ExecutionWorker {
     pub fn new(
-        cfg: ExecutionConfig,
+        cfg: ExecutionSettings,
         thread_count: u8,
         event_sender: mpsc::UnboundedSender<ExecutionEvent>,
         controller_command_rx: mpsc::Receiver<ExecutionCommand>,
         controller_manager_rx: mpsc::Receiver<ExecutionManagementCommand>,
+        bootstrap_state: Option<BootstrapExecutionState>,
     ) -> Result<ExecutionWorker, ExecutionError> {
         let execution_queue = ExecutionQueue::default();
         let execution_queue_clone = execution_queue.clone();
         let cfg_clone = cfg.clone();
-        // Start vm thread
+
+        // Check bootstrap
+        let bootstrap_final_slot;
+        let bootstrap_ledger;
+        if let Some(bootstrap_state) = bootstrap_state {
+            // init from bootstrap
+            bootstrap_final_slot = bootstrap_state.final_slot;
+            bootstrap_ledger = Some((bootstrap_state.final_ledger, bootstrap_final_slot));
+        } else {
+            // init without bootstrap
+            bootstrap_final_slot = Slot::new(0, thread_count.saturating_sub(1));
+            bootstrap_ledger = None;
+        };
+
+        // Init VM
+        let vm = Arc::new(Mutex::new(VM::new(cfg, thread_count, bootstrap_ledger)?));
+        let vm_clone = vm.clone();
```

Comment:
Why do you clone ?

Reply:
Here you just increment the atomical reference counter as a shared_pointer. It's because vm object is used in two OS threads

Reply:
because vm_thread takes a clone, and ExecutionWorker keeps the original